File: python/src/kanbus/issue_transfer.py
"""Local and shared issue transfer helpers."""

from __future__ import annotations

from pathlib import Path

from kanbus.issue_files import read_issue_from_file
from kanbus.models import IssueData
from kanbus.project import (
    ProjectMarkerError,
    ensure_project_local_directory,
    find_project_local_directory,
    load_project_directory,
)
from kanbus.event_history import (
    create_event,
    events_dir_for_local,
    events_dir_for_project,
    now_timestamp,
    transfer_payload,
    write_events_batch,
)
from kanbus.users import get_current_user


class IssueTransferError(RuntimeError):
    """Raised when moving issues between shared and local storage fails."""


def promote_issue(root: Path, identifier: str) -> IssueData:
    """Move a local issue into the shared project directory.

    :param root: Repository root path.
    :type root: Path
    :param identifier: Issue identifier.
    :type identifier: str
    :return: Issue data that was promoted.
    :rtype: IssueData
    :raises IssueTransferError: If promotion fails.
    """
    try:
        project_dir = load_project_directory(root)
    except ProjectMarkerError as error:
        raise IssueTransferError(str(error)) from error

    local_dir = find_project_local_directory(project_dir)
    if local_dir is None:
        raise IssueTransferError("project-local not initialized")

    local_issue_path = local_dir / "issues" / f"{identifier}.json"
    if not local_issue_path.exists():
        raise IssueTransferError("not found")

    target_path = project_dir / "issues" / f"{identifier}.json"
    if target_path.exists():
        raise IssueTransferError("already exists")

    issue = read_issue_from_file(local_issue_path)
    local_issue_path.replace(target_path)
    occurred_at = now_timestamp()
    actor_id = get_current_user()
    event = create_event(
        issue_id=issue.identifier,
        event_type="issue_promoted",
        actor_id=actor_id,
        payload=transfer_payload("local", "shared"),
        occurred_at=occurred_at,
    )
    try:
        events_dir = events_dir_for_project(project_dir)
        write_events_batch(events_dir, [event])
    except Exception as error:  # noqa: BLE001
        target_path.replace(local_issue_path)
        raise IssueTransferError(str(error)) from error
    return issue


def localize_issue(root: Path, identifier: str) -> IssueData:
    """Move a shared issue into the project-local directory.

    :param root: Repository root path.
    :type root: Path
    :param identifier: Issue identifier.
    :type identifier: str
    :return: Issue data that was localized.
    :rtype: IssueData
    :raises IssueTransferError: If localization fails.
    """
    try:
        project_dir = load_project_directory(root)
    except ProjectMarkerError as error:
        raise IssueTransferError(str(error)) from error

    shared_issue_path = project_dir / "issues" / f"{identifier}.json"
    if not shared_issue_path.exists():
        raise IssueTransferError("not found")

    local_dir = ensure_project_local_directory(project_dir)
    target_path = local_dir / "issues" / f"{identifier}.json"
    if target_path.exists():
        raise IssueTransferError("already exists")

    issue = read_issue_from_file(shared_issue_path)
    shared_issue_path.replace(target_path)
    occurred_at = now_timestamp()
    actor_id = get_current_user()
    event = create_event(
        issue_id=issue.identifier,
        event_type="issue_localized",
        actor_id=actor_id,
        payload=transfer_payload("shared", "local"),
        occurred_at=occurred_at,
    )
    try:
        events_dir = events_dir_for_local(project_dir)
        write_events_batch(events_dir, [event])
    except Exception as error:  # noqa: BLE001
        target_path.replace(shared_issue_path)
        raise IssueTransferError(str(error)) from error
    return issue

File: rust/src/issue_transfer.rs
//! Local and shared issue transfer helpers.

use std::fs;
use std::path::Path;

use crate::error::KanbusError;
use crate::event_history::{
    events_dir_for_local, events_dir_for_project, now_timestamp, transfer_payload, write_events_batch,
    EventRecord, EventType,
};
use crate::file_io::{
    ensure_project_local_directory, find_project_local_directory, load_project_directory,
};
use crate::issue_files::read_issue_from_file;
use crate::models::IssueData;
use crate::users::get_current_user;

/// Promote a local issue into the shared project directory.
///
/// # Arguments
/// * `root` - Repository root path.
/// * `identifier` - Issue identifier.
///
/// # Errors
/// Returns `KanbusError::IssueOperation` if promotion fails.
pub fn promote_issue(root: &Path, identifier: &str) -> Result<IssueData, KanbusError> {
    let project_dir = load_project_directory(root)?;
    let local_dir = find_project_local_directory(&project_dir)
        .ok_or_else(|| KanbusError::IssueOperation("project-local not initialized".to_string()))?;

    let local_issue_path = local_dir.join("issues").join(format!("{identifier}.json"));
    if !local_issue_path.exists() {
        return Err(KanbusError::IssueOperation("not found".to_string()));
    }

    let target_path = project_dir
        .join("issues")
        .join(format!("{identifier}.json"));
    if target_path.exists() {
        return Err(KanbusError::IssueOperation("already exists".to_string()));
    }

    let issue = read_issue_from_file(&local_issue_path)?;
    fs::rename(&local_issue_path, &target_path)
        .map_err(|error| KanbusError::Io(error.to_string()))?;

    let occurred_at = now_timestamp();
    let actor_id = get_current_user();
    let event = EventRecord::new(
        issue.identifier.clone(),
        EventType::IssuePromoted,
        actor_id,
        transfer_payload("local", "shared"),
        occurred_at,
    );
    let events_dir = events_dir_for_project(&project_dir);
    match write_events_batch(&events_dir, &[event]) {
        Ok(_paths) => {}
        Err(error) => {
            fs::rename(&target_path, &local_issue_path)
                .map_err(|io_error| KanbusError::Io(io_error.to_string()))?;
            return Err(error);
        }
    }
    Ok(issue)
}

/// Move a shared issue into the project-local directory.
///
/// # Arguments
/// * `root` - Repository root path.
/// * `identifier` - Issue identifier.
///
/// # Errors
/// Returns `KanbusError::IssueOperation` if localization fails.
pub fn localize_issue(root: &Path, identifier: &str) -> Result<IssueData, KanbusError> {
    let project_dir = load_project_directory(root)?;
    let shared_issue_path = project_dir
        .join("issues")
        .join(format!("{identifier}.json"));
    if !shared_issue_path.exists() {
        return Err(KanbusError::IssueOperation("not found".to_string()));
    }

    let local_dir = ensure_project_local_directory(&project_dir)?;
    let target_path = local_dir.join("issues").join(format!("{identifier}.json"));
    if target_path.exists() {
        return Err(KanbusError::IssueOperation("already exists".to_string()));
    }

    let issue = read_issue_from_file(&shared_issue_path)?;
    fs::rename(&shared_issue_path, &target_path)
        .map_err(|error| KanbusError::Io(error.to_string()))?;

    let occurred_at = now_timestamp();
    let actor_id = get_current_user();
    let event = EventRecord::new(
        issue.identifier.clone(),
        EventType::IssueLocalized,
        actor_id,
        transfer_payload("shared", "local"),
        occurred_at,
    );
    let events_dir = match events_dir_for_local(&project_dir) {
        Ok(path) => path,
        Err(error) => {
            fs::rename(&target_path, &shared_issue_path)
                .map_err(|io_error| KanbusError::Io(io_error.to_string()))?;
            return Err(error);
        }
    };
    match write_events_batch(&events_dir, &[event]) {
        Ok(_paths) => {}
        Err(error) => {
            fs::rename(&target_path, &shared_issue_path)
                .map_err(|io_error| KanbusError::Io(io_error.to_string()))?;
            return Err(error);
        }
    }
    Ok(issue)
}
